// ironworks/src/eventing/CommEvent.js
var __extends = this.__extends || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    __.prototype = b.prototype;
    d.prototype = new __();
};
var _ = require('lodash');
var Event = require('./Event');
var CommEvent = (function (_super) {
    __extends(CommEvent, _super);
    function CommEvent(event) {
        var evt;
        if (typeof event === 'string') {
            evt = {
                prefix: void 0,
                service: void 0,
                method: void 0,
                worker: void 0,
                name: event
            };
        }
        else {
            evt = event;
        }
        var split = evt.name.split('.').reverse();
        evt.prefix = _.isUndefined(split[4]) ? evt.prefix : split[4];
        evt.service = _.isUndefined(split[3]) ? evt.service : split[3];
        evt.method = _.isUndefined(split[2]) ? evt.method : split[2];
        evt.worker = _.isUndefined(split[1]) ? evt.worker : split[1];
        evt.name = _.isUndefined(split[0]) ? evt.name : split[0];
        _super.call(this, evt);
        this.prefix = evt.prefix;
        this.service = evt.service;
        this.method = evt.method;
        this.worker = evt.worker;
    }
    CommEvent.prototype.getText = function () {
        return _.compact([
            this.prefix,
            this.service,
            this.method,
            this.worker,
            this.name
        ]).join('.');
    };
    CommEvent.prototype.equal = function (event) {
        var prefixMatch = this.prefix === '*' || event.prefix === '*' || this.prefix === event.prefix;
        var serviceMatch = this.service === '*' || event.service === '*' || this.service === event.service;
        var methodMatch = this.method === '*' || event.method === '*' || this.method === event.method;
        var workerMatch = this.worker === '*' || event.worker === '*' || this.worker === event.worker;
        var nameMatch = this.name === '*' || event.name === '*' || this.name === event.name;
        return prefixMatch && serviceMatch && methodMatch && workerMatch && nameMatch;
    };
    return CommEvent;
})(Event);
module.exports = CommEvent;
